### PR TITLE
Introduce "per task source" cancellation

### LINKED ISSUE #21119: TaskSource-specific TaskCanceller
In order to support stuff like https://github.com/servo/servo/issues/21114, we need the ability to cancel tasks in a more granular way, at the level of 'per tasksource'. 

Currently 'cancelling' tasks is basically dependent on the [ignore_further_async_events](https://github.com/servo/servo/blob/1f9b07637ea960f6b3d4004f832951ad9b6b48ed/components/script/dom/window.rs#L251) field on `Window`, which is the basis for the `TaskCanceller` returned via [task_canceller](https://github.com/servo/servo/blob/1f9b07637ea960f6b3d4004f832951ad9b6b48ed/components/script/dom/window.rs#L1068), which is accessed via the [queue](https://github.com/servo/servo/blob/master/components/script/task_source/mod.rs#L29) method of the [TaskSource](https://github.com/servo/servo/blob/master/components/script/task_source/mod.rs#L16) trait(via the globalscope passed to that function). 

One idea I can come up with is adding a new method to `TaskSource`, something like
`fn choose_canceller<T>(&self, global: &GlobalScope) -> TaskCanceller`, and then change `queue` to read something like the below:
```
let canceller = self.choose_canceller(&global);
self.queue_with_canceller(task, canceller);
```

Then each task source will need to implement `choose_canceller`, calling the appropriate method of `GlobalScope`, which would forward the call to the window, and the window would require a flag similar to [ignore_further_async_events](https://github.com/servo/servo/blob/1f9b07637ea960f6b3d4004f832951ad9b6b48ed/components/script/dom/window.rs#L251) for each task source...

Any better ideas?

### PR DESCRIPTION
**Pull request by gterzian**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #21119 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21126)

### CODE REVIEW

**Comment by gterzian** [comment]
This one is ready for review, although I'm wondering about the right testing strategy for this one...

**Comment by gterzian** [comment]
For clarity this isn't suppose to 'change' anything about the behavior of the system, it only lays the groundwork to start changing behavior, for example in https://github.com/servo/servo/issues/21114, which will require using a dedicated task source for `EventSource` and then only cancelling the `networking tasks` as part of [document.abort](https://github.com/servo/servo/blob/0125e96555d86e27097172fe6132f32fcf3c8870/components/script/dom/document.rs#L2034), and letting the event source tasks run. 

So perhaps a good way to test is 'making sure everything still works as before', and then it is the follow up work that will include tests specifically asserting that certain tasks are cancelled while others still run?

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/layout_image.rs [hunk 1]
```diff
@@ -62,7 +63,7 @@ pub fn fetch_image_for_layout(url: ServoUrl,
     let listener = NetworkListener {
         context: context,
         task_source: window.networking_task_source(),
-        canceller: Some(window.task_canceller()),
+        canceller: Some(window.task_canceller(TaskSourceName::Networking)),
```
[the remark this answers is not in the document]
Comment (gterzian) [reply components/script/layout_image.rs]: I've filed https://github.com/servo/servo/issues/21154

**Comment by gterzian** [comment]
@jdm Thanks for the pointers, comments have been addressed...

**Pull request merged**